lib/client/ipam_test.go

```diff
@@ -0,0 +1,300 @@
+// - Assign 1 address on host A (Expect 1 address)
+// - Assign 1 address on host B (Expect 1 address, different block)
+// - Assign 64 more addresses on host A (Expect 63 addresses from host A's block, 1 address from host B's block)
+
+// Test cases (AssignIP):
+// Test 1: Assign 1 IPv4 from a configured pool - expect no error returned.
+// Test 2: Assign 1 IPv6 from a configured pool - expect no error returned.
+// Test 3: Assign 1 IPv4 from a non-configured pool - expect an error returned.
+// Test 4: Assign 1 IPv4 from a configured pool twice:
+// - Expect no error returned while assigning the IP for the first time.
+// - Expect an error returned while assigning the SAME IP again.
+
+// Test cases (ReleaseIPs):
+// Test 1: Assign 1 IPv4 with AssignIP from a configured pool and then release it.
+// - Assign should not return an error.
+// - ReleaseIPs should return empty slice of IPs (unallocatedIPs) and no error.
+// Test 2: Assign 66 IPs (across 2 blocks) with AutoAssign from a configured pool then release them.
+// - Assign should not return an error.
+// - ReleaseIPs should return an empty slice of IPs (unallocatedIPs) and no error.
+// Test 3: Assign 1 IPv4 address with AssignIP then try to release 2 IPs.
+// - Assign should not return an error.
+// - ReleaseIPs should return a slice with one (unallocatedIPs) and an error???
+// Test 4: release an IP that's not configured in any pools - expect nil slice??? and an error.
+// Test 5: release an IP that's not allocated in the pool - expect a slice with one (unallocatedIPs) and an error.
+
+// Test cases (ClaimAffinity):
+// Test 1: claim affinity for an unclaimed IPNet of size 64 - expect the same IPNet and empty slice of IPNet with no error.
+// Test 2: claim affinity for a IPNet that has an IP already assigned to another host.
+// - Assign an IP with AssignIP to "host-A" from a configured pool - expect no error.
+// - Claim affinity to the block that IP belongs to - expect ????.
+// Test 3: claim affinity to a block twice.
+// - Claim affinity to an unclaimed block - expect to get the same IPNet back in claimed IP slice and no error.
+// - Claim affinity to the same block again - expect the IPNet back as the second return value (failed) and an error.
+
+package client_test
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"net"
+	"os/exec"
+	"strings"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	. "github.com/onsi/ginkgo/extensions/table"
+
+	"github.com/tigera/libcalico-go/calicoctl/commands"
+	"github.com/tigera/libcalico-go/lib/api"
+	"github.com/tigera/libcalico-go/lib/backend/etcd"
+	"github.com/tigera/libcalico-go/lib/client"
+	cnet "github.com/tigera/libcalico-go/lib/net"
+)
+
+// Setting BackendType to etcdv2 which is the only supported backend at the moment.
+var etcdType api.BackendType = "etcdv2"
+
+// Setting localhost as the etcd endpoint location since that's where `make run-etcd` runs it.
+var etcdConfig = etcd.EtcdConfig{
+	EtcdEndpoints: "http://127.0.0.1:2379",
```

Comment:
Ditto, it doesn't work with empty config, that is it gets empty string as EtcdEndpoints instead of going default, which I think is because the the default is set for when decoding JSON:
 `json:"etcdAuthority" envconfig:"ETCD_AUTHORITY" default:"127.0.0.1:2379"`